File: elixir/triangle/triangle.exs
defmodule Triangle do
  @type kind :: :equilateral | :isosceles | :scalene

  @doc """
  Return the kind of triangle of a triangle with 'a', 'b' and 'c' as lengths.
  """
  @spec kind(number, number, number) :: { :ok, kind } | { :error, String.t }
  def kind(a, b, c) when a <= 0 or b <= 0 or c <= 0 do
    { :error, "all side lengths must be positive" }
  end
  def kind(a, b, c) when a >= b + c or b >= a + c or c >= a + b do
    { :error, "side lengths violate triangle inequality" }
  end
  def kind(a, a, a), do: { :ok, :equilateral }
  def kind(a, a, _), do: { :ok, :isosceles }
  def kind(a, _, a), do: { :ok, :isosceles }
  def kind(_, a, a), do: { :ok, :isosceles }
  def kind(a, b, c), do: { :ok, :scalene }
end

File: ecmascript/triangle/triangle.js
class Triangle {
  constructor(...sides) {
    this.sides = sides;
  }

  kind() {
    if (this.isInvalid()) throw new Error('Not a triangle');
    if (this.isEquilateral()) return 'equilateral';
    if (this.isIsosceles()) return 'isosceles';
    return 'scalene';
  }

  isInvalid() {
    if (!this.validSides() || !this.satisfyTriangleInequality())
      return true;
    return false;
  }

  validSides() {
    return this.sides.every(x => x > 0);
  }

  satisfyTriangleInequality() {
    const [a, b, c] = this.sides;
    return a+b>=c && a+c>=b && b+c>=a;
  }

  isEquilateral() {
    return new Set(this.sides).size == 1;
  }

  isIsosceles() {
    return new Set(this.sides).size == 2;
  }
}

export default Triangle;
